airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/ctid/InitialSyncCtidIterator.java

```diff
@@ -0,0 +1,230 @@
+  private final CtidPostgresSourceOperations sourceOperations;
+  private final Queue<Pair<Ctid, Ctid>> subQueriesPlan;
+  private final String tableName;
+  private final long tableSize;
+  private AutoCloseableIterator<RowDataWithCtid> currentIterator;
+
+  private Long lastKnownFileNode;
+  private int numberOfTimesReSynced = 0;
+
+  public InitialSyncCtidIterator(final CtidStateManager ctidStateManager,
+      final JdbcDatabase database,
+      final CtidPostgresSourceOperations sourceOperations,
+      final String quoteString,
+      final List<String> columnNames,
+      final String schemaName,
+      final String tableName,
+      final long tableSize,
+      final long blockSize,
+      final FileNodeHandler fileNodeHandler) {
+    this.airbyteStream = AirbyteStreamUtils.convertFromNameAndNamespace(tableName, schemaName);
+    this.blockSize = blockSize;
+    this.columnNames = columnNames;
+    this.ctidStateManager = ctidStateManager;
+    this.database = database;
+    this.fileNodeHandler = fileNodeHandler;
+    this.quoteString = quoteString;
+    this.schemaName = schemaName;
+    this.sourceOperations = sourceOperations;
+    this.subQueriesPlan = new LinkedList<>();
+    this.tableName = tableName;
+    this.tableSize = tableSize;
+  }
+
+  @CheckForNull
+  @Override
+  protected RowDataWithCtid computeNext() {
+    try {
+      if (subQueriesPlan.isEmpty() && currentIterator == null) {
+        initSubQueries();
+      }
+
+      if (currentIterator == null || !currentIterator.hasNext()) {
+        final Optional<Long> mayBeLatestFileNode = PostgresQueryUtils.fileNodeForIndividualStream(database, airbyteStream, quoteString);
+        if (mayBeLatestFileNode.isPresent()) {
+          final Long latestFileNode = mayBeLatestFileNode.get();
+          if (lastKnownFileNode != null) {
+            if (!latestFileNode.equals(lastKnownFileNode)) {
+              resetSubQueries(latestFileNode);
+            } else {
+              LOGGER.info("The latest file node {} for stream {} is equal to the last known file node {} known to Airbyte.",
+                  latestFileNode,
+                  lastKnownFileNode,
+                  airbyteStream);
+            }
+          }
+          lastKnownFileNode = latestFileNode;
+          fileNodeHandler.updateFileNode(airbyteStream, latestFileNode);
+        } else {
+          LOGGER.warn("Airbyte could not query the latest file node for stream {}. Continuing sync as usual.", airbyteStream);
+        }
+
+        if (currentIterator != null) {
+          currentIterator.close();
+        }
+
+        if (subQueriesPlan.isEmpty()) {
+          return endOfData();
+        }
+
+        final Pair<Ctid, Ctid> p = subQueriesPlan.remove();
+        currentIterator = AutoCloseableIterators.fromStream(getStream(p), airbyteStream);
+      }
+
+      if (currentIterator.hasNext()) {
+        return currentIterator.next();
+      }
+
+      assert subQueriesPlan.isEmpty();
```

Comment:
Can we always assert that this is true? 

Consider : we have pre-computed 10 subqueries, each corresponding to a range of Ctids. I think it's completely plausible for some range in the middle (e.g. 6th query) to not return any data. In this case - we'll fail this assertion. 

In fact - we should keep going until subQueriesPlan is empty to prevent this situation